[DRAFT] 2.0 upgrade guide working doc (#2298)
microsoft/OpenAPI.NET: docs/upgrade-guide-2.md

# Performance Improvements

One of the key features of OpenAPI.NET is its performance. This version makes it possible to parse JSON based OpenAPI descriptions even faster. OpenAPI.NET v1 relied on the excellent YamlSharp library for parsing both JSON and YAML files. With OpenAPI.NET v2 we are relying on System.Text.Json for parsing JSON files. For YAML files, we continue to use YamlSharp to parse YAML but then convert to JsonNodes for processing. This allows us to take advantage of the performance improvements in System.Text.Json while still supporting YAML files.

In v1, instances of `$ref` were resolved in a second pass of the document to ensure the target of the reference has been parsed before attempting to resolve it. In v2, reference targets are lazily resolved when reference objects are accessed. This improves load time performance for documents that make heavy use of references.

[How does this change the behaviour of external references?]

## Reduced Dependencies

> In OpenAPI v1, it was necessary to include the Microsoft.OpenApi.Readers library to be able to read OpenAPI descriptions in either YAML or JSON.  In OpenAPI.NET v2, the core Microsoft.OpenAPI library can both read and write JSON.  It is only necessary to use the newly renamed [Microsoft.OpenApi.YamlReader](https://www.nuget.org/packages/Microsoft.OpenApi.YamlReader/) library if you need YAML support. This allows teams who are only working in JSON to avoid the additional dependency and therefore eliminate all non-.NET library references.

Reviewer: A code snippet of how to do this would be useful. Presumably:

```csharp
var settings = new OpenApiReaderSettings();
settings.AddYamlReader();

var result = OpenApiDocument.LoadAsync(openApiString, settings: settings);
```

Reviewer 2: fixed, thanks

In v1, instances of `$ref` were resolved in a second pass of the document to ensure the target of the reference has been parsed before attempting to resolve it. In v2, reference targets are lazily resolved when reference objects are accessed. This improves load time performance for documents that make heavy use of references.

[How does this change the behaviour of external references?]

## Reduced Dependencies

In OpenAPI v1, it was necessary to include the Microsoft.OpenApi.Readers library to be able to read OpenAPI descriptions in either YAML or JSON.  In OpenAPI.NET v2, the core Microsoft.OpenAPI library can both read and write JSON.  It is only necessary to use the newly renamed [Microsoft.OpenApi.YamlReader](https://www.nuget.org/packages/Microsoft.OpenApi.YamlReader/) library if you need YAML support. This allows teams who are only working in JSON to avoid the additional dependency and therefore eliminate all non-.NET library references.

## API Enhancements

> The v1 library attempted to mimic the pattern of `XmlTextReader` and `JsonTextReader` for the purpose of loading OpenAPI documents from strings, streams and textReaders.

Reviewer: ```suggestion
The v1 library attempted to mimic the pattern of `XmlTextReader` and `JsonTextReader` for the purpose of loading OpenAPI documents from strings, streams and text readers.
```

```csharp
    var reader = new OpenApiStringReader();
    var openApiDoc = reader.Read(stringOpenApiDoc, out var diagnostic);
```
The same pattern can be used for `OpenApiStreamReader` and `OpenApiTextReader`.  When we introduced the `ReadAsync` methods we eliminated the use of the `out` parameter.

```csharp
    var reader = new OpenApiStreamReader();
    var (document, diagnostics) = await reader.ReadAsync(streamOpenApiDoc);
```
A `ReadResult` object acts as a tuple of `OpenApiDocument` and `OpenApiDiagnostic`.

> The challenge with this approach is that the reader classes are not very discoverable and the behaviour is not actually consistent with `*TextReader` pattern that allows incrementally reading the document. This library does not support incrementally reading the OpenAPI Document. It only reads a complete document and returns an `OpenApiDocument` instance.

Reviewer: ```suggestion
The challenge with this approach is that the reader classes are not very discoverable and the behaviour is not actually consistent with the `*TextReader` pattern that allows incrementally reading the document. This library does not support incrementally reading the OpenAPI Document. It only reads a complete document and returns an `OpenApiDocument` instance.
```

In the v2 library we are moving to the pattern used by classes like `XDocument` where a set of static `Load` and `Parse` methods are used as factory methods.

```csharp
public class OpenApiDocument {
    public static async Task<ReadResult> LoadAsync(string url, OpenApiReaderSettings settings = null) {}
    public static async Task<ReadResult> LoadAsync(Stream stream, string? format = null, OpenApiReaderSettings? settings = null) {}
    public static ReadResult Load(MemoryStream stream, string? format = null, OpenApiReaderSettings? settings = null) {}

>     public static ReadResult Parse(string input,string? format = null,OpenApiReaderSettings? settings = null) {}

Reviewer: ```suggestion
    public static ReadResult Parse(string input, string? format = null, OpenApiReaderSettings? settings = null) {}
```

}
```

> This API design allows a developer to use IDE autocomplete to present all the loading options by simply knowning the name of the `OpenApiDocument` class.  Each of these methods are layered ontop of the more primitive methods to ensure consistent behaviour.

Reviewer: ```suggestion
This API design allows a developer to use IDE autocomplete to present all the loading options by simply knowing the name of the `OpenApiDocument` class.  Each of these methods are layered on top of the more primitive methods to ensure consistent behaviour.
```

> As the YAML format is only supported when including the `Microsoft.OpenApi.YamlReader` library it was decided not to use an enum for the `format` parameter.  We are considering implementing a more [strongly type solution](https://github.com/microsoft/OpenAPI.NET/issues/1952) similar to the way that `HttpMethod` is implemented so that we have a strongly typed experience that is also extensible.

Reviewer: ```suggestion
As the YAML format is only supported when including the `Microsoft.OpenApi.YamlReader` library it was decided not to use an enum for the `format` parameter.  We are considering implementing a more [strongly typed solution](https://github.com/microsoft/OpenAPI.NET/issues/1952) similar to the way that `HttpMethod` is implemented so that we have a strongly typed experience that is also extensible.
```

> Where the loading methods are used without a format property, we will attempt to parse the document using the default JSON reader.  If that fails and the Yaml reader is registered, then we will attempt to read as YAML.  The goal is always to provide the fastest path with JSON but still maintain the convenience of not having to care whether a URL points to YAML or JSON if you need that flexibility.

Reviewer: ```suggestion
When the loading methods are used without a format parameter, we will attempt to parse the document using the default JSON reader.  If that fails and the YAML reader is registered, then we will attempt to read as YAML.  The goal is always to provide the fastest path with JSON but still maintain the convenience of not having to care whether a URL points to YAML or JSON if you need that flexibility.
```

### Removing the OpenAPI Any classes

In the OpenAPI specification, there are a few properties that are defined as type `any`. This includes:
- the example property in the parameter, media type objects
- the value property in the example object

> - the values in the link object's parameters dictionary and requestBody property

Reviewer: ```suggestion
- the values in the link object's parameters dictionary and `requestBody` property
```

- all `x-` extension properties

> In the v1 library, there are set of classes that are derived from the OpenApiAny base class which an abstract model which reflects the JSON data model plus some additional primitive types such as decimal, float, datetime etc.

Reviewer: ```suggestion
In the v1 library, there are a set of classes that are derived from the `OpenApiAny` base class which is an abstract model which reflects the JSON data model plus some additional primitive types such as `decimal`, `float`, `datetime` etc.
```

> In v2 we are removing this abstraction and relying on the JsonNode model to represent these inner types. In v1 we were not able to reliably identify the additional primitive types and it caused a significant amount of false negatives in error reporting as well as incorrectly parsed data values.

Reviewer: ```suggestion
In v2 we are removing this abstraction and relying on the `JsonNode` model to represent these inner types. In v1 we were not able to reliably identify the additional primitive types and it caused a significant amount of false negatives in error reporting as well as incorrectly parsed data values.
```

> Due to JsonNode implicit operators, this makes initialization sometimes easier, instead of:

Reviewer: ```suggestion
Due to `JsonNode` implicit operators, this makes initialization sometimes easier, instead of:
```

```csharp
    new OpenApiParameter
                {
                    In = null,
                    Name = "username",
                    Description = "username to fetch",
                    Example = new OpenApiFloat(5),
                };
```
the assignment becomes simply,
```csharp

>                     Example = (float)0.5,

Reviewer: ```suggestion
                    Example = 0.5f,
```

        public string Comment { get; set; }
        /// $vocabulary- used in meta-schemas to identify the vocabularies available for use in schemas described by that meta-schema.
        public IDictionary<string, bool> Vocabulary { get; set; }
        /// $dynamicRef - an applicator that allows for deferring the full resolution until runtime, at which point it is resolved each time it is encountered while evaluating an instance
        public string DynamicRef { get; set; }
        /// $dynamicAnchor - used to create plain name fragments that are not tied to any particular structural location for referencing purposes, which are taken into consideration for dynamic referencing.
        public string DynamicAnchor { get; set; }
        /// $defs - reserves a location for schema authors to inline re-usable JSON Schemas into a more general schema.
        public IDictionary<string, OpenApiSchema> Definitions { get; set; }
        public IDictionary<string, OpenApiSchema> PatternProperties { get; set; } = new Dictionary<string, OpenApiSchema>();
        public bool UnevaluatedProperties { get; set;}

>         public bool UnEvaluatedProperties { get; set; }  // Duplicate should be removed

Reviewer: ?

```suggestion
```

```
#### Changes to existing keywords

```csharp

        public virtual decimal? ExclusiveMaximum { get; set; }  // New, but currently named v31ExclusiveMaximum
        public virtual decimal? ExclusiveMinimum { get; set; } // New, but Currently named v31ExclusiveMinimum)
        public virtual decimal? ExclusiveMaximum { get; set; }  //type changed to reflect the new version of JSON schema

>         public virtual decimal? ExclusiveMinimum { get; set; } // type changed to reflect the new version of JSON schema

Reviewer: These are duplicated.

        public virtual JsonSchemaType? Type { get; set; }  // Was string, now flagged enum
        public virtual decimal? Maximum { get; set; }      // Double???

>         public virtual decimal? Minimum { get; set; }       // Double???

Reviewer: I assume these are TODO comments?

Reviewer 2: fixed, thanks

        public virtual JsonNode Default { get; set; }  // Type matching no longer enforced. Was IOpenApiAny
        public virtual bool ReadOnly { get; set; }  // No longer has defined semantics in OpenAPI 3.1
        public virtual bool WriteOnly { get; set; }  // No longer has defined semantics in OpenAPI 3.1

        public virtual JsonNode Example { get; set; }  // No longer IOpenApiAny
        public virtual IList<JsonNode> Examples { get; set; }

>         public virtual IList<JsonNode> Enum { get; set; } = new List<JsonNode>();

Reviewer: ```suggestion
        public virtual IList<JsonNode> Enum { get; set; }
```

        public virtual OpenApiExternalDocs ExternalDocs { get; set; }  // OpenApi Vocab
        public virtual bool Deprecated { get; set; }  // OpenApi Vocab
        public virtual OpenApiXml Xml { get; set; }  // OpenApi Vocab

>         public IDictionary<string, object> Annotations { get; set; }  // Custom keywords?

Reviewer: ?

Reviewer 2: fixed, thanks

```

#### OpenApiSchema methods

> Other than the addition of the SerializeAsV31, the methods have not changed.

Reviewer: ```suggestion
Other than the addition of `SerializeAsV31`, the methods have not changed.
```

```csharp
public class OpenApiSchema : IOpenApiAnnotatable, IOpenApiExtensible, IOpenApiReferenceable, IOpenApiSerializable
{
        public OpenApiSchema() { }
        public OpenApiSchema(OpenApiSchema schema) { }
        public virtual void SerializeAsV31(IOpenApiWriter writer) { }
        public virtual void SerializeAsV3(IOpenApiWriter writer) { }
        public virtual void SerializeAsV2(IOpenApiWriter writer) { }
}

```

## OpenAPI v3.1 Support

> There are a number of new features in OpenAPI v3.1 that are now supported in OpenAPI.Net.

Reviewer: ```suggestion
There are a number of new features in OpenAPI v3.1 that are now supported in OpenAPI.NET.
```

### Webhooks

```csharp

public class OpenApiDocument  : IOpenApiSerializable, IOpenApiExtensible, IOpenApiAnnotatable {
        /// <summary>
        /// The incoming webhooks that MAY be received as part of this API and that the API consumer MAY choose to implement.
        /// A map of requests initiated other than by an API call, for example by an out of band registration. 
        /// The key name is a unique string to refer to each webhook, while the (optionally referenced) Path Item Object describes a request that may be initiated by the API provider and the expected responses
        /// </summary>
        public IDictionary<string, OpenApiPathItem>? Webhooks { get; set; } = new Dictionary<string, OpenApiPathItem>();

> 

Reviewer: ```suggestion
```

}
```

### Summary in info object

```csharp

    /// <summary>
    /// Open API Info Object, it provides the metadata about the Open API.
    /// </summary>
    public class OpenApiInfo : IOpenApiSerializable, IOpenApiExtensible
    {
        /// <summary>
        /// A short summary of the API.
        /// </summary>
        public string Summary { get; set; }
    }
```

### License SPDX identifiers

```csharp

> /// <summary>

Reviewer: ```suggestion
    /// <summary>
```

    /// License Object.
    /// </summary>
    public class OpenApiLicense : IOpenApiSerializable, IOpenApiExtensible
    {
        /// <summary>

>         /// An SPDX license expression for the API. The identifier field is mutually exclusive of the url field.

Reviewer: ```suggestion
        /// An SPDX license expression for the API. The identifier field is mutually exclusive of the Url property.
```

        /// </summary>
        public string Identifier { get; set; }
    }
```

### Reusable path items

```csharp
    /// <summary>
    /// Components Object.
    /// </summary>
    public class OpenApiComponents : IOpenApiSerializable, IOpenApiExtensible
    {
        /// <summary>
        /// An object to hold reusable <see cref="OpenApiPathItem"/> Object.
        /// </summary>
        public virtual IDictionary<string, OpenApiPathItem>? PathItems { get; set; } = new Dictionary<string, OpenApiPathItem>();

> 

Reviewer: ```suggestion
```

    }
```

#### Summary and Description alongside $ref

Through the use of proxy objects in order to represent references, it is now possible to set the Summary and Description property on an object that is a reference. This was previously not possible.

```csharp
    var parameter = new OpenApiParameterReference("id", hostdocument)
    {
        Description = "Customer Id"
    };
```
### Use HTTP Method Object Instead of Enum
HTTP methods are now represented as objects instead of enums. This change enhances flexibility but requires updates to how HTTP methods are handled in your code.
Example:
```csharp
// Before (1.6)
OpenApiOperation operation = new OpenApiOperation
{
    HttpMethod = OperationType.Get
};

// After (2.0)
OpenApiOperation operation = new OpenApiOperation
{

>     HttpMethod = new HttpMethod("GET")// or HttpMethod.Get

Reviewer: ```suggestion
    HttpMethod = new HttpMethod("GET") // or HttpMethod.Get
```
